GitHub actions, black format, drop python2

jccurtis opened the pull request:
The individual changes can be viewed in the commit messages. Here is a summary:

CI:

- [x] Add basic GH actions test workflow
    - [x] Make required for merge in the [branch protection rules](https://github.com/lbl-anp/becquerel/settings/branches)
- [x] Remove travis CI 
    - [x] Remove config
    - [x] Remove travis trigger [here](https://travis-ci.org/organizations/lbl-anp/repositories)
    - [x] Disable webhook [here](https://github.com/lbl-anp/becquerel/settings/hooks)
    - [x] Remove required for merge in the [branch protection rules](https://github.com/lbl-anp/becquerel/settings/branches)

Formatting:

- [x] Format with `black`

Drop python2

- [x] `from __future__`
- [x] `sys.version_info`
- [x] `ImportError` checks

Misc

- [x] Remove `pylint` dep and disable comments (@markbandstra lemme know if you want to revert)

Future work:

- Python3.9 support: #243

Closes #216, #125, #126, #220, #221, #233, #182

jccurtis commented:
@markbandstra @jvavrek @cosama I think this is good for a first pass. Comments?

jccurtis commented:
The testing OS/version converage is more comprehensive but runs in less time than on travis 👍

markbandstra commented on becquerel/core/rebin.py:
```diff
@@ -315,11 +324,14 @@ def rebin(in_spectra, in_edges, out_edges, method="interpolation",
-            raise RebinError('Cannot rebin spectra with negative values with '
-                             'listmode method')
+            raise RebinError(
+                "Cannot rebin spectra with negative values with " "listmode method"
```
Can combine these into one string

jccurtis replied:
Fixed: 89b19f071234854759897f5697ca87b8a20439ff

markbandstra commented on becquerel/tools/nndc.py:
```diff
@@ -226,65 +229,63 @@ def _parse_float_uncertainty(x, dx):
-        raise NNDCRequestError(
-            'Value cannot be parsed as float: "{}"'.format(x))
-    if dx == '':
+        raise NNDCRequestError('Value cannot be parsed as float: "{}"'.format(x))
```
Is it better to leave strings like these with single quotes if they enclose double quotes, or to make them doubles and escape the inner double quotes?

cosama replied:
```suggestion
        raise NNDCRequestError(f'Value cannot be parsed as float: "{x}"')
```
Need to move to the future :smile: No strong feelings about quotes. Think this is fine.

jccurtis replied:
Cleaned. b13c0aa52cd68cf402792f049bae90ab5096bf48 I'm good with the current formatting but can change. I think that is outside of this PR eh?

markbandstra replied:
Agreed, we can change it later if it bothers us.

markbandstra commented on pyproject.toml:
```diff
@@ -0,0 +1,16 @@
+[tool.black]
+target-version = ['py36', 'py37', 'py38']
```
Also `py39`?

jccurtis replied:
See comment in description. I had the same thought but my dreams were crushed by `llvmlite` 😉

markbandstra commented on setup.py:
```diff
@@ -2,17 +2,23 @@
 """Becquerel: Tools for radiation spectral analysis."""
 
 from __future__ import print_function
```
We can now safely remove all `from __future__ import` statements.

markbandstra replied:
Oops, looks like you took care of this right before my comment.

cosama commented on .github/workflows/pip_build_test.yaml:
```diff
@@ -0,0 +1,43 @@
+    # OS and python combos to test
+    strategy:
+      matrix:
+        os: [ 'macos-latest', 'ubuntu-latest']
```
windows?

jccurtis replied:
Good point. Here is the [OS listing](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources). Updated: 91df59000e5beb821aab244439abd1deb0b8ebcd

<img width="567" alt="image" src="https://user-images.githubusercontent.com/1504659/110147860-00146a80-7d91-11eb-8784-626fcc24643c.png">

cosama replied:
looks good. If it fails on windows, I would create a issue and take it off.

jccurtis commented:
@markbandstra Can you look at the failing tests? I'm not sure if it is NNDC access or I broke a string format somewhere ...

jvavrek commented on CONTRIBUTING.md:
```diff
@@ -47,19 +47,16 @@ Use [google standards](https://google.github.io/styleguide/pyguide.html)
+We like to use [`flake8`](https://flake8.pycqa.org/en/latest/) and/or [`pylance`](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance).
 
 ### Checklist for code contributions:
   - [ ] Branch off of `main` and name the branch `feature-XX` or `issue-XX`
```
We haven't been too strict on this

jccurtis replied:
Yeah, I'm good removing. @markbandstra @cosama ?

markbandstra replied:
I'm fine with removal.

cosama replied:
me too.

jccurtis replied:
Done

jccurtis commented:
Actually everything is working now. Including the windows testing 😊

jvavrek commented on becquerel/core/rebin.py:
```diff
@@ -55,14 +54,16 @@ def _check_partial_overlap(in_edges, out_edges):
+            + "zeros will padded on the left side of the new spectrum",
+            RebinWarning,
+        )
     if (in_edges[..., -1] < out_edges[..., -1]).any():
         warnings.warn(
-            'The last input edge is smaller than the last output edge, ' +
-            'zeros will padded on the right side of the new spectrum',
-            RebinWarning)
+            "The last input edge is smaller than the last output edge, "
+            + "zeros will padded on the right side of the new spectrum",
```
The `+` are unnecessary here, right?

markbandstra commented:
I didn't get to look at it, any idea what the problem was?

jccurtis commented:
@markbandstra it looked like two things 

1. A failure in accessing NNDC (random)
2. A string formatting bug I introduced but just patched.

jvavrek commented on becquerel/parsers/spectrum_file.py:
```diff
@@ -66,46 +66,50 @@ def __str__(self):
-        warnings.warn('The use of energy_bin_edges is deprecated, ' +
-                      'use bin_edges_kev instead', DeprecationWarning)
+        warnings.warn(
+            "The use of energy_bin_edges is deprecated, " + "use bin_edges_kev instead",
```
could be cleaned up

jccurtis commented:
@jvavrek any further thoughts?

jvavrek commented on becquerel/tools/df_cache.py:
```diff
@@ -59,10 +57,9 @@ def check_path(self):
 
         if not os.path.exists(self.path):
-            raise CacheError('Cache path does not exist: {}'.format(self.path))
+            raise CacheError("Cache path does not exist: {}".format(self.path))
```
All of these could be upgraded to f-strings

jvavrek commented on setup.py:
```diff
@@ -23,58 +28,61 @@
 License :: Other/Proprietary License
```
Is this correct?

markbandstra replied:
Yes, as I understand it, it means we have a custom license, not BSD/MIT/GPL/etc.

jvavrek reviewed (commented):
> @jvavrek any further thoughts?

I have a few small review comments. Just going through the rest now.

jvavrek commented on becquerel/core/spectrum.py:
```diff
@@ -146,21 +155,23 @@ def __init__(self, counts=None, cps=None, uncs=None, bin_edges_kev=None,
-                    'are most likely not Poisson-distributed. Provide uncs ' +
-                    'to force initialization.')
+                    "Negative values encountered in counts. Uncertainties "
+                    + "are most likely not Poisson-distributed. Provide uncs "
```
remove `+` ?

jvavrek commented on becquerel/core/spectrum.py:
```diff
@@ -941,14 +980,15 @@ def has_uniform_bins(self, use_kev=None, rtol=None):
-            raise UncalibratedError('Cannot access energy bins with an ' +
-                                    'uncalibrated Spectrum.')
+            raise UncalibratedError(
+                "Cannot access energy bins with an " + "uncalibrated Spectrum."
```
remove `+`

jvavrek commented on becquerel/core/spectrum.py:
```diff
@@ -991,18 +1031,19 @@ def find_bin_index(self, x, use_kev=None):
-            raise UncalibratedError('Cannot access energy bins with an ' +
-                                    'uncalibrated Spectrum.')
+            raise UncalibratedError(
+                "Cannot access energy bins with an " + "uncalibrated Spectrum."
```
remove `+`

jvavrek commented on becquerel/core/spectrum.py:
```diff
@@ -1134,27 +1180,37 @@ def rebin(self, out_edges, method="interpolation", slopes=None,
-            raise SpectrumError('Cannot rebin spectrum without energy '
-                                'calibration')  # TODO: why not?
+            raise SpectrumError(
+                "Cannot rebin spectrum without energy " "calibration"
```
concatenate

jvavrek commented on tests/xcom_test.py:
```diff
@@ -1,33 +1,123 @@
+ENERGIES_3 = [60.0, 662.0, 1460.0]
 
 
 # standard grid energies for Germanium from 1 keV to 10 MeV
```
this is unfortunate. maybe worth #noqa if that still works with black?

markbandstra replied:
Not sure what you mean. Because they aren't the actual line energies anymore with one more sig fig?

jvavrek replied:
Oh, I meant to point out the code below "standard grid energies" is now a long list of one entry per line. You do raise a good point though.

markbandstra replied:
Oh, I see. I'm not worried about those numbers being the right energies. Yes, the formatting of the lists below are much less compact, but I don't think we can do anything about that.

jvavrek commented on tests/nndc_test.py:
```diff
@@ -31,114 +27,114 @@ class TestParseFloatUncertainty(object):
 
     def test_01(self):
         """Test _parse_float_uncertainty('257.123', '0.005')..............."""
```
are all the `.......` necessary?

markbandstra replied:
Haha, no, that is a very old hack from me to get the output from `nose` to look pretty and aligned when there are a lot of tests. Not part of this PR, but can be removed in the future.

jvavrek commented on becquerel/core/plotting.py:
```diff
@@ -484,12 +473,13 @@ def ylim(self):
-            elif yscale == 'symlog' and data_min >= 0:
+            elif yscale == "log" and data_min < 0:
+                raise PlottingError(
+                    "Cannot plot negative values on a log " + "scale; use symlog scale"
```
concatenate

jvavrek reviewed (commented):
added a few format suggestions, but looks good otherwise

jccurtis commented:
Let's finish this up on Monday. I will get to the comments by then.

jccurtis commented:
> > @jvavrek any further thoughts?
> 
> I have a few small review comments. Just going through the rest now.

@jvavrek I addressed the comments. I have not updated all the `f` strings because there are 100's and I think we could leave the rest for a future MR which will also help catching copy/paste and other errors.

Only thing left is the unresolved thread above.

jccurtis commented:
Okay everything is resolved. Once the tests pass can I get a final review?

markbandstra reviewed (approved):
@jccurtis Well, it was a slog, but it looks like you've done it!
